theano/tensor/alt_gemm_template.c

```diff
@@ -0,0 +1,133 @@
+/** %(name)s **/
+void alt_numpy_scalar_matrix_product_in_place_%(float_type)s(%(float_type)s scalar, PyArrayObject* matrix) {
+    NpyIter* iterator = NpyIter_New(matrix, 
+        NPY_ITER_READWRITE | NPY_ITER_EXTERNAL_LOOP | NPY_ITER_REFS_OK, 
+        NPY_KEEPORDER, NPY_NO_CASTING, NULL);
+    if(iterator == NULL)
+        alt_fatal_error("Unable to iterate over a matrix "
+                        "for a scalar * matrix operation.");
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterator, NULL);
+    char** data_ptr = NpyIter_GetDataPtrArray(iterator);
+    npy_intp* stride_ptr = NpyIter_GetInnerStrideArray(iterator);
+    npy_intp* innersize_ptr = NpyIter_GetInnerLoopSizePtr(iterator);
+    do {
+        char* data = *data_ptr;
+        npy_intp stride = *stride_ptr;
+        npy_intp count = *innersize_ptr;
+        while(count) {
+            *((%(float_type)s*)data) *= scalar;
+            data += stride;
+            --count;
+        }
+    } while(get_next(iterator));
+    NpyIter_Deallocate(iterator);
+}
+/*Matrix+Matrix function. Compute (coeffA * matrixA) + (coeffB * matrixB)
+ * Remark: This function actually sums a C-contiguous matrix (alpha*op(A)*op(B)) with a F-contiguous matrix (beta*C)
+ * (see gemm implementation at next function for more details) */
+void alt_numpy_matrix_extended_sum_in_place_%(float_type)s(
+        const %(float_type)s* ALPHA, PyArrayObject* A,
+        const %(float_type)s* BETA, PyArrayObject* B
+) {
+    PyArrayObject* op[2]       = {A, B};
+    npy_uint32     op_flags[2] = {NPY_ITER_READONLY, NPY_ITER_READWRITE};
+    npy_uint32     flags       = 0;
+    NpyIter*       iterators   = NpyIter_MultiNew(
+            2, op, flags, NPY_CORDER, NPY_NO_CASTING, op_flags, NULL);
+    if(iterators == NULL)
+        alt_fatal_error("Unable to iterate over some matrices "
+                        "for matrix + matrix operation.");
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterators, NULL);
+    char** data_ptr_array = NpyIter_GetDataPtrArray(iterators);
+    do {
+        %(float_type)s* from_A = (%(float_type)s*)data_ptr_array[0];
+        %(float_type)s* from_B = (%(float_type)s*)data_ptr_array[1];
+        *from_B = (*ALPHA)*(*from_A) + (*BETA)*(*from_B);
+    } while(get_next(iterators));
+    NpyIter_Deallocate(iterators);
+}
+PyObject* alt_op_without_copy_%(float_type)s(int transposable, %(float_type)s* M, int nrow, int ncol, int LDM) {
+    // By default, M is considered as a nrow*ncol F-contiguous matrix with LDM as stride indicator for the columns.
+    npy_intp dims[2];
+    npy_intp strides[2];
+    int flags;
+    if(transposable) {
+        dims[0] = ncol;
+        dims[1] = nrow;
+        strides[0] = dims[1] * %(float_size)d;
+        strides[1] = %(float_size)d;
+        flags = NPY_ARRAY_C_CONTIGUOUS;
```

Comment:
OK, I think I misunderstood the doc for `PyArray_New()`, indeed there will not be a copy.
However, I think the `CONTIGUOUS` flags can be too string (the array is not considered C-contiguous if `LDM != dims[1]`, or F-contiguous if `LDM != dims[0]`, even though it is good enough for BLAS).
I think we may need to manually test corner cases.